### DISCUSSION TITLE
ChanQueueSubscribe slow consumer help needed

### DISCUSSION THREAD

**Discussion by susana-garcia** [nats-io/nats.go#1084] [General]
Hello,

We are having an issue with our nats consumers and I’m wondering if someone could help me. We monitor our NATS JetStream with prometheus/grafana and there I see:
* 12.8K pending messages and there are increasing by a 1000 per hour since a few hours.
* Messages added and processed per second per consumer vary between 0.3 and 1.

We use `ChanQueueSubscribe` and I saw sometimes in the logs a `nats: slow consumer, messages dropped` error message. But weird enough, I see new messages coming into both running apps and doing what they’re supposed to do. Plus it takes a few milliseconds to handle a new message received by the consumer.

Any idea what might be off? 

We use for the consumers and this is our setup:
<img width="440" alt="Screenshot 2022-09-15 at 15 38 22" src="https://user-images.githubusercontent.com/10615081/190595141-59a67910-beb4-48b6-b063-b47abac084d0.png">
<img width="781" alt="Screenshot 2022-09-15 at 15 38 51" src="https://user-images.githubusercontent.com/10615081/190595151-bac5fb37-0918-47ef-ae9f-c2c0c099c96f.png">

Thank you in advance.

**Comment by wallyqs** [comment]
What is the size of the channel? It needs to have enough capacity to handle the traffic, something like 512k cap is the default in other APIs like `Subscribe`.

**Comment by susana-garcia** [comment]
Uhi, that might be the problem then, the size is 1 😳 
I'll update it and let you know if it helped.
Thank you very much for the support.

**Comment by wallyqs** [comment]
no problem! :) will move to discussion instead of issue

**Reply by susana-garcia** [reply]
Hi again,
Hope you had a nice weekend.

I haven’t done the channel capacity change yet, because I’d like to first understand how the current setup is working:
- The messaging system seems to be working as expected so far: messages are published correctly, consumed and updates are made (based on the consumed messages). 
- Yesterday looking at the logs, I saw that there were around 17K pending messages in grafana, but there’re no new ones being published at the time and logs were static (we have debug mode to see when a new message is published and received by the consumers, nothing was happening at the time). Oddly enough the pending messages weren't decreasing either. 
- At some point, grafana shows that the pending messages dropped to zero.

Questions: 
1. Why do pending messages increase usually exponentially? We use `LimitsPolicy` as `RetentionPolicy` with `MaxAge` 7 days and `MaxBytes` 1 GiB, but we also have a `AckExplicitPolicy` where we `Ack` or `NakWithDelay` messages. When a message is consumed and we acknowledge it, that message should be removed, right? Also when a message is negatively acknowledge, it stays in the queue until the `MaxDeliver` (3 in our case) is reached, correct? In the logs I see that the last few days, we only `NakWithDelay` around 200 messages and `Ack` around 27K and none of these requests (NakWithDelay/Ack) failed, based on the logs.
2. Why are the messages being removed suddenly? I guess it’s because one of MaxAge/MaxBytes were reached.
3. If we increase the capacity of the channel and the application suddenly shuts down, the messages that are in the channel at that moment, will get lost or as we don’t NakWithDelay/Ack them, will be re deliver?
4. Do you know if it’s possible to see the pending messages and their content using nats cli?

Regards,
Susana

**Reply by wallyqs** [reply]
Hi @susana-garcia could you share a bit more about the environment like server and client version? Are you using the prometheus nats exporter/nats-surveyor in the setup?
The num pending increasing if there are no new messages being produced sounds like would be due to redeliveries happening. The messages that are in the Go channel are in 'delivered but waiting for an ack`'/ num_waiting, so if the application shuts down without draining the channel they would remain inflight until being redelivered.

**Reply by susana-garcia** [reply]
Of course, sorry, I forgot to share more of the environment before:
```
github.com/nats-io/nats-server/v2 v2.8.2
github.com/nats-io/nats.go v1.16.0
```
We use argocd and the nats server helm chart version is `2.8.4`
For monitoring we use: https://github.com/nats-io/prometheus-nats-exporter/tree/main/walkthrough

Please, let me know if you need more info.
And thank you for clarifying what happens to the messages in the go channel.

**Reply by wallyqs** [reply]
Thanks @susana-garcia for the info, one first recommendation would be to try to upgrade to the v2.9.0 which includes a lot of fixes.

**Reply by susana-garcia** [reply]
ok, thanks, I'll update it too.

**Reply by susana-garcia** [reply]
hi @wallyqs 
We've updated nats server version to `2.9.0` and increase the capacity of the channel and so far so good.
Thank you for all.
Regards,
Susana

**Reply by susana-garcia** [reply]
@wallyqs I believe there is something else going on. The updates of the new events keep arriving in the consumers punctually. Could it be that the already ack messages are not being deleted from the stream somehow? 
<img width="412" alt="Screenshot 2022-09-23 at 14 08 18" src="https://user-images.githubusercontent.com/10615081/191957143-78b75830-3010-4fc9-ae59-8d6b8f6161bb.png">

**Reply by susana-garcia** [reply]
Also, shouldn't be the pending messages seen here in `Unprocessed Messages` (`nats consumer info`) or that's something else?
<img width="751" alt="Screenshot 2022-09-23 at 15 37 51" src="https://user-images.githubusercontent.com/10615081/191972827-17f76e69-b0e2-48e5-a945-103f5762d212.png">

**Reply by susana-garcia** [reply]
So, I've realized I didn't understand how the stream works with different retention policies and consumers.
I made a POC and here is what I found out: I tried Interest and Limits Retention Policies with push consumers and Interest Retention Policy with pull consumer.
The only one that deletes the messages after ack them is Interest Retention Policy with push consumer.

**Reply by ripienaar** [reply]
Work queue policy deletes after any 1 ack from any 1 consumer (and so overlaps in interest is not allowed)

interest policy deletes after all consumers with matching interest acked a message.

**Reply by susana-garcia** [reply]
Exactly, also I forgot to mention, but with limits, the messages get deleted when one of the [limits defined](https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive) is reached. Sorry, it's my bad I assumed that that happened automatically after ack a message (in the case of Limits Retention Policy).
Thanks for all.

**Reply by ripienaar** [reply]
yes, ilmits apply to all cases regardless of retention method, they work together.
